controller/controller.go

```diff
@@ -261,42 +263,50 @@ func (c *Controller) propagate(resourceEvents []ResourceEvent) {
 	}
 }
 
-func (c *Controller) subscribe() {
-	cache, ok := c.cache.(*watchableCacheStore) // should we add Subscribe(ctx) to the Cache interface or remove the interface altogether?
-	if !ok {
-		return
-	}
-	recent := make(Store)
-	subscription := cache.Subscribe(context.TODO())
+func (c *Controller) subscribe(ctx context.Context) {
+	oldObjs := make(Store)
+	subscription := c.cache.SubscribeSubset(ctx, func(storeId string, _ Store) bool {
+		return storeId == resourceStoreId
+	})
 	go func() {
 		for snapshot := range subscription {
 			c.Lock()
 
-			c.propagate(lo.FlatMap(snapshot.Updates, func(update watchable.Update[string, watchableCacheEntry], _ int) []ResourceEvent {
-				key := update.Key
-				obj := update.Value
+			newObjs := snapshot.State[resourceStoreId]
 
+			events := lo.FilterMap(lo.Keys(newObjs), func(uid string, _ int) (ResourceEvent, bool) {
+				newObj := newObjs[uid]
 				event := ResourceEvent{
-					Kind: obj.GetObjectKind().GroupVersionKind().GroupKind(),
+					Kind:      newObj.GetObjectKind().GroupVersionKind().GroupKind(),
+					NewObject: newObj,
+				}
+				if oldObj, exists := oldObjs[uid]; !exists {
+					event.EventType = CreateEvent
+					oldObjs[uid] = newObj
+					return event, true
+				} else if !reflect.DeepEqual(oldObj, newObj) {
+					event.EventType = UpdateEvent
+					event.OldObject = oldObj
+					oldObjs[uid] = newObj
+					return event, true
 				}
+				return event, false
+			})
 
-				if update.Delete {
-					event.EventType = DeleteEvent
-					event.OldObject = obj
-					delete(recent, key)
-				} else {
-					if oldObj, ok := recent[key]; ok {
-						event.EventType = UpdateEvent
-						event.OldObject = oldObj
-					} else {
-						event.EventType = CreateEvent
-					}
-					event.NewObject = obj
-					recent[key] = obj
+			deleteEvents := lo.FilterMap(lo.Keys(oldObjs), func(uid string, _ int) (ResourceEvent, bool) {
+				oldObj := oldObjs[uid]
+				event := ResourceEvent{
+					EventType: DeleteEvent,
+					Kind:      oldObj.GetObjectKind().GroupVersionKind().GroupKind(),
+					OldObject: oldObj,
 				}
+				_, exists := newObjs[uid]
+				return event, !exists
```

Comment:
```suggestion
				_, exists := newObjs[uid]
				if !exists {
					delete(oldObjs, uid)
					return event, true
				}
				return event, false
```

Delete entry from `oldObjs` for delete event?

Reply:
😵‍💫

Reply:
Fixed.